### PR TITLE
feat: Report histogram metrics to Triton metrics server

### PR DESCRIPTION
**Pull request by yinggeh**
Sample histogram output
```
# HELP vllm:time_to_first_token_seconds Histogram of time to first token in seconds.
# TYPE vllm:time_to_first_token_seconds histogram
vllm:time_to_first_token_seconds_count{model="vllm_opt",version="1"} 3
vllm:time_to_first_token_seconds_sum{model="vllm_opt",version="1"} 0.0002238750457763672
vllm:time_to_first_token_seconds_bucket{model="vllm_opt",version="1",le="0.001"} 3
vllm:time_to_first_token_seconds_bucket{model="vllm_opt",version="1",le="0.005"} 3
vllm:time_to_first_token_seconds_bucket{model="vllm_opt",version="1",le="0.01"} 3
vllm:time_to_first_token_seconds_bucket{model="vllm_opt",version="1",le="0.02"} 3
vllm:time_to_first_token_seconds_bucket{model="vllm_opt",version="1",le="0.04"} 3
vllm:time_to_first_token_seconds_bucket{model="vllm_opt",version="1",le="0.06"} 3
vllm:time_to_first_token_seconds_bucket{model="vllm_opt",version="1",le="0.08"} 3
vllm:time_to_first_token_seconds_bucket{model="vllm_opt",version="1",le="0.1"} 3
vllm:time_to_first_token_seconds_bucket{model="vllm_opt",version="1",le="0.25"} 3
vllm:time_to_first_token_seconds_bucket{model="vllm_opt",version="1",le="0.5"} 3
vllm:time_to_first_token_seconds_bucket{model="vllm_opt",version="1",le="0.75"} 3
vllm:time_to_first_token_seconds_bucket{model="vllm_opt",version="1",le="1"} 3
vllm:time_to_first_token_seconds_bucket{model="vllm_opt",version="1",le="2.5"} 3
vllm:time_to_first_token_seconds_bucket{model="vllm_opt",version="1",le="5"} 3
vllm:time_to_first_token_seconds_bucket{model="vllm_opt",version="1",le="7.5"} 3
vllm:time_to_first_token_seconds_bucket{model="vllm_opt",version="1",le="10"} 3
vllm:time_to_first_token_seconds_bucket{model="vllm_opt",version="1",le="+Inf"} 3
# HELP vllm:time_per_output_token_seconds Histogram of time per output token in seconds.
# TYPE vllm:time_per_output_token_seconds histogram
vllm:time_per_output_token_seconds_count{model="vllm_opt",version="1"} 45
vllm:time_per_output_token_seconds_sum{model="vllm_opt",version="1"} 0.002027750015258789
vllm:time_per_output_token_seconds_bucket{model="vllm_opt",version="1",le="0.01"} 45
vllm:time_per_output_token_seconds_bucket{model="vllm_opt",version="1",le="0.025"} 45
vllm:time_per_output_token_seconds_bucket{model="vllm_opt",version="1",le="0.05"} 45
vllm:time_per_output_token_seconds_bucket{model="vllm_opt",version="1",le="0.075"} 45
vllm:time_per_output_token_seconds_bucket{model="vllm_opt",version="1",le="0.1"} 45
vllm:time_per_output_token_seconds_bucket{model="vllm_opt",version="1",le="0.15"} 45
vllm:time_per_output_token_seconds_bucket{model="vllm_opt",version="1",le="0.2"} 45
vllm:time_per_output_token_seconds_bucket{model="vllm_opt",version="1",le="0.3"} 45
vllm:time_per_output_token_seconds_bucket{model="vllm_opt",version="1",le="0.4"} 45
vllm:time_per_output_token_seconds_bucket{model="vllm_opt",version="1",le="0.5"} 45
vllm:time_per_output_token_seconds_bucket{model="vllm_opt",version="1",le="0.75"} 45
vllm:time_per_output_token_seconds_bucket{model="vllm_opt",version="1",le="1"} 45
vllm:time_per_output_token_seconds_bucket{model="vllm_opt",version="1",le="2.5"} 45
vllm:time_per_output_token_seconds_bucket{model="vllm_opt",version="1",le="+Inf"} 45
```

#### What does the PR do?
Support histogram metric type and add tests.

#### Checklist
- [x] PR title reflects the change and is of format `<commit_type>: <Title>`
- [x] Changes are described in the pull request.
- [x] Related issues are referenced.
- [x] Populated [github labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels) field
- [x] Added [test plan](#test-plan) and verified test passes.
- [x] Verified that the PR passes existing CI.
- [x] Verified copyright is correct on all changed files.
- [x] Added _succinct_ git squash message before merging [ref](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
- [x] All template sections are filled out.
- [ ] Optional: Additional screenshots for behavior/output changes with before/after.

#### Commit Type:
Check the [conventional commit type](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type)
box here and add the label to the github PR.
- [x] feat

#### Related PRs:
https://github.com/triton-inference-server/python_backend/pull/374
https://github.com/triton-inference-server/core/pull/386
https://github.com/triton-inference-server/server/pull/7525

#### Where should the reviewer start?
n/a

#### Test plan:
n/a

- CI Pipeline ID:
17487728

#### Caveats:
n/a

#### Background
Customer requested histogram metrics from vLLM.

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)
n/a

### CODE REVIEW

**Comment by oandreeva-nv** [comment]
Could you possibly target the initial metrics branch `yinggeh-DLIS-7061-add-vllm-metrics`? It would be easier to review this way

File: src/model.py
Comment (oandreeva-nv) [inline src/model.py:163]: Question for this and `gauge/counter` PR as well. Would this cause trouble if `tritonserver` was built without metrics ?

File: src/utils/metrics.py
Comment (oandreeva-nv) [inline src/utils/metrics.py:71]: Could you please clarify why we're using these numbers?

File: src/model.py
Comment (GuanLuo) [reply src/model.py:163]: @rmccorm4 how will the backend APIs behave ^^

File: src/utils/metrics.py
Comment (yinggeh) [reply src/utils/metrics.py:55]: @oandreeva-nv Explanation to [comment](https://github.com/triton-inference-server/vllm_backend/pull/56#discussion_r1707859240).

File: src/utils/metrics.py
Comment (yinggeh) [reply src/utils/metrics.py:71]: Buckets here are just example from vLLM repo [metrics.py](https://github.com/vllm-project/vllm/blob/main/vllm/engine/metrics.py#L81-L96).

Comment (yinggeh) [reply src/utils/metrics.py:71]: Buckets here are just example from vLLM repo [metrics.py](https://github.com/vllm-project/vllm/blob/main/vllm/engine/metrics.py#L81-L96).

File: src/utils/metrics.py
Comment (oandreeva-nv) [reply src/utils/metrics.py:55]: I think, if we ship `utils/metrics.py` as a part of a supported backend, we need to ship a defined set of buckets anyways, at least as a default. Since we ship this as a python script, users can always adjust it on their side. Since these values correspond to vllm side of things, I think it worth adding a comment about it with a permalink, so that we could easily refer to the original source and adjust

File: src/model.py
Comment (oandreeva-nv) [reply src/model.py:163]: My concern comes from here: https://github.com/triton-inference-server/core/pull/386/files#diff-5ed8dbbcfda4302eb8115993e5297cdf1e184f856c415059946525a7b618397cR3461-R3464
If metrics are not enabled, we send an error. IIUC this and `gauge` PR will make vllm backend use metrics all the time?

File: ci/L0_backend_vllm/metrics_test/vllm_metrics_test.py
Comment (rmccorm4) [reply ci/L0_backend_vllm/metrics_test/vllm_metrics_test.py:44]: This env var is used to work for windows tests as well because localhost doesn't currently work on our windows tests (cc @fpetrini) but agree the hard-coded cases should be swapped to use shared variable

**Comment by yinggeh** [comment]
> Could you possibly target the initial metrics branch `yinggeh-DLIS-7061-add-vllm-metrics`? It would be easier to review this way

@oandreeva-nv Target?

File: src/model.py
Comment (yinggeh) [reply src/model.py:163]: Please use this [discussion](https://github.com/triton-inference-server/vllm_backend/pull/53#discussion_r1697679955) to track the issue.

**Comment by oandreeva-nv** [comment]
@yinggeh , i.e. instead of merging this branch into main, set the target merge branch as `yinggeh-DLIS-7061-add-vllm-metrics` , this PR significantly overlaps with `gauge/counter` metrics PR, so it would be easier to review

**Comment by yinggeh** [comment]
> @yinggeh , i.e. instead of merging this branch into main, set the target merge branch as `yinggeh-DLIS-7061-add-vllm-metrics` , this PR significantly overlaps with `gauge/counter` metrics PR, so it would be easier to review

Is it better now?

File: ci/L0_backend_vllm/metrics_test/vllm_metrics_test.py
Comment (yinggeh) [reply ci/L0_backend_vllm/metrics_test/vllm_metrics_test.py:73]: Is there a performance boost otherwise?

Comment (GuanLuo) [reply ci/L0_backend_vllm/metrics_test/vllm_metrics_test.py:73]: The goal of the test is to test metrics, infer is not the main point here. This is just a helper function to run some inference and async / stream just complicates things

**Comment by oandreeva-nv** [comment]
@yinggeh I think you would need to resolve conflicts as well. Basically, it is still better to re-base this branch on top of `yinggeh-DLIS-7061-add-vllm-metrics`. If not now, then when `yinggeh-DLIS-7061-add-vllm-metrics` is merged to main, same conflicts will be present.

File: ci/L0_backend_vllm/metrics_test/vllm_metrics_test.py
Comment (oandreeva-nv) [reply ci/L0_backend_vllm/metrics_test/vllm_metrics_test.py:73]: I had same question [here](https://github.com/triton-inference-server/vllm_backend/pull/53#discussion_r1704522802). Just send a simple request to the model

**Comment by yinggeh** [comment]
> @yinggeh I think you would need to resolve conflicts as well. Basically, it is still better to re-base this branch on top of `yinggeh-DLIS-7061-add-vllm-metrics`. If not now, then when `yinggeh-DLIS-7061-add-vllm-metrics` is merged to main, same conflicts will be present.

Do I need to resolve conflicts for `yinggeh-DLIS-7113-support-histogram-metric-type` each time `yinggeh-DLIS-7061-add-vllm-metrics` has a new commit?

**Comment by oandreeva-nv** [comment]
Depending on the commit.

File: ci/L0_backend_vllm/metrics_test/vllm_metrics_test.py
Comment (yinggeh) [reply ci/L0_backend_vllm/metrics_test/vllm_metrics_test.py:73]: client.infer won't work.
`I0808 23:09:24.212543 85125 infer_handler.cc:975] "[request id: <id_unknown>] Infer failed: ModelInfer RPC doesn't support models with decoupled transaction policy"`

**Comment by yinggeh** [comment]
> Depending on the commit.

Rebased.

File: src/utils/metrics.py
Comment (yinggeh) [reply src/utils/metrics.py:73]: Updated

File: README.md
Comment (GuanLuo) [inline README.md:262]:
```suggestion
...
```

**Review by GuanLuo — APPROVED** [review]

File: README.md
Comment (yinggeh) [reply README.md]: Updated

**Comment by rmccorm4** [comment]
Note: new commits should dismiss approvals, but it looks like that setting wasn't applied in this repo -- will update the settings, but please keep that in mind :pray:

**Comment by yinggeh** [comment]
@rmccorm4 Thanks. Btw I merged to the wrong branch... It was initially set to merge to `yinggeh-DLIS-7061-add-vllm-metrics` to help reviewers skipped duplicate changes. I am working on new PRs to merge them into main.

**Comment by fpetrini15** [comment]
@yinggeh @rmccorm4 Repo rules have been updated to dismiss stale approvals when new commits are pushed. In the future, please do let the operations team know if the repo rules appear to be incorrect.

**Pull request merged**
